### PR TITLE
fix(curriculum): Added a Recursion challenge to the Basic JavaScript section

### LINKED ISSUE #16107: Recursion all but required for many Intermediate Algorithms, but it has not yet been introduced
#### Challenge Name
https://www.freecodecamp.org/challenges/steamroller

Hint 3 states: "You will definitely need recursion...." However, there has not yet been any mention of recursion in the lessons. Recursion needs to be taught (or at the very least mentioned) in the lessons no later than Steamroller.

(The last few lessons have been grueling to solve without recursion.)

### PR DESCRIPTION
**Pull request by RandellDawson**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #16107

I made this PR after looking at #35966.  While that PR does introduce a challenge which uses recursion, I felt the example and the actual challenge needed to be more similar in nature, so as not to be too big of a leap for a camper's first introduction to writing a recursive function.  I did borrow the other PR's first sentence in the Description "Recursion is the programming concept of using a function that calls itself in order to accomplish a task." but everything else is much different.

I want some feedback regarding how the Description section could be improved.

### CODE REVIEW

**Comment by dnntrn** [comment]
If this is the learner's first introduction to recursion, I still think they might struggle to understand. I personally struggled a lot with understanding return statements in recursive functions, and didn't really understand until it was visualized through a recursion tree. Otherwise, I think the description is very clear!

Is there a way for us to either include a some sort of visualization to demonstrate what the return statements do or maybe link to a recursion visualizer of some sort?

**Comment by RandellDawson** [comment]
> Is there a way for us to either include a some sort of visualization to demonstrate what the return statements do or maybe link to a recursion visualizer of some sort?

If the visualization can be done without images.  Remember, we have people using screen readers, so an image would not help them.

**Comment by dnntrn** [comment]
Not sure if this is good, but just as an example of what I'm thinking of that wouldn't require images:

To understand the return statements in recursion functions, it is helpful to think of the
recursive call as "pausing" the function for that value.

For an example, consider countToN(3).

At first, n = 3.
....So, var numbers = countToN(2) is called and the function with n = 3 is "paused" until a return statement is reached
....Then, n = 2
........So, var numbers = countToN(1) is called and the function with n = 2 is "paused" until a return statement is reached
........Then, n = 1
............So, [1] is returned. The return statement is reached, and the function resumes the above calls.

....var numbers now equals [1] (due to the return statement) and n = 2 again
........Since n is pushed to numbers, numbers now equals [1,2]. A return statement is reached again.
var numbers now equals [1,2] (due to the return statement) and n = 3 again
....Since n is pushed to numbers, numbers now equals [1,2,3]. A return statement is reached again, and all calls have completed.

This with proper indentation:

Screen Shot 2019-05-10 at 8 59 39 PM

Or some other type of formatting that makes it clear the general structure of recursive functions and it "jumping back" after a return statement is hit.

**Comment by RandellDawson** [comment]
That is good information and might be better place in the Guide, so that when a user clicks the Get a Hint button, they see this information first.  I think adding all of that to the challenge will make it too much text and too much to read through.  We want the user to be able to read through a challenge description and instructions in 2 minutes or less.  I think this much text would be too much.  It may be we need to have a really basic recursive challenge first and then one like mine above. 

Did you take a look at the other PR's (https://github.com/freeCodeCamp/freeCodeCamp/pull/35966) approach for a recursive challenge?  Maybe that one could be tweaked a bit to make it a little more beginner friendly and then use mine as a more advanced challenge?  Feel free to make further recommendations on the other PR, this PR, or the original issue (#16107).

**Comment by ojeytonwilliams** [comment]
I agree that this level of detail would be make a nice guide article, but would be a bit much for a challenge.  Given that recursion is a tricky concept to grasp, it would be good to have more than one challenge dealing with it.  So, yes, I like the idea of this being an advanced/practice challenge following up on #35966.

One thing I thought was, could we do something even simpler that the recursion in this PR or #35966?  Maybe a sum:

```js
function sum(xs, n) {
  if (n <= 0) {
    return xs[0];
  } else {
    return sum(xs, n-1) + xs[n];
  }
}
```

That should be easier to understand because you can express it like this:

```
sum(xs, 2) = sum(xs, 1) + xs[2] 
           = (sum(xs, 0) + xs[1]) + xs[2]
           = (xs[0] + xs[1]) + xs[2]
           = xs[0] + xs[1] + xs[2]
```
and you don't need to care that `sum(xs,0)` returns first, since order doesn't matter for addition.  Finally, nothing gets mutated, so it's easier to reason about.

I'm not sure where this would fit, though.  Perhaps it would be best as the first introduction, followed by #35966 and finally this PR.  @RandellDawson @dnntrn what do you think?

**Comment by RandellDawson** [comment]
@ojeytonwilliams I think your idea could work.  Create a PR and we can look at it more.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/using-recursion-instead-of-loops.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/using-recursion-instead-of-loops.english.md]:
```suggestion
We have defined a function named <code>rangeOfNumbers</code> with two parameters. The function should return an array of integers which begins with a starting number represented by the <code>startNum</code> parameter and ends with an ending number represented by the <code>endNum</code> parameter.  The starting number will always be less than or equal to the ending number.  Your function must use recursion by calling itself and not use loops of any kind.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/using-recursion-instead-of-loops.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/using-recursion-instead-of-loops.english.md]:
```suggestion
    testString: assert(removeJSComments(rangeOfNumbers.toString()).match(/rangeOfNumbers.toString()).match(/rangeOfNumbers\(.*\).*\{.*rangeOfNumbers\(.*\).*\}/s));
```
I noticed when making my recursion challenge that this test passes with an empty function, like
```js
function rangeOfNumbers(startNum, endNum) {

};
```
so I came up with this.

**Review by ojeytonwilliams — DISMISSED** [review]
Couple of minor things, otherwise LGTM.  Ideally we should coordinate with @rns350 and try and make our lessons work as beginning, intermediate and advanced. Assuming we want all three.  As the latecomer, obviously mine would go if we decide on only having two.

**Comment by moT01** [comment]
So, are we still wanting to add this since the other lesson got merged?

**Comment by RandellDawson** [comment]
@moT01 I was thinking we were still considering this PR and https://github.com/freeCodeCamp/freeCodeCamp/pull/35966 as part of a 3 part series of recursive challenges.  @ojeytonwilliams Is that not what you were thinking?

**Comment by ojeytonwilliams** [comment]
That was what I had in mind.  My PR was only ever supposed to be a very gentle introduction, since recursion can be a tricky concept to grasp.  It isn't enough on its own.

**Comment by moT01** [comment]
the challenge title for the merged PR and the challenge title on this PR are very similar - you three creating these PR's should maybe come up with a little outline or something so there's a little progression of difficulty or something - maybe you already have something in mind. I'll leave it to you guys for now - let me know if you need anything.

**Comment by ojeytonwilliams** [comment]
My PR might have been merged a little early, but I'd be happy to revisit it to make them fit better.  Any ideas on how to progress in difficulty would definitely help.

**Comment by moT01** [comment]
I haven't reviewed these PR's, so I'm not sure what could be added or modified without looking through them. But, you would think to start with the most basic example and move to something a little more realistic/difficult.

**Comment by RandellDawson** [comment]
@ojeytonwilliams The single backticks work fine for the `<code>` tags, so that is good.  Thanks for your help on making the wording for this challenge.  Just need to review/approve and post in the Gitter room.

Thanks.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]:
```suggestion
We have defined a function named <code>rangeOfNumbers</code> with two parameters. The function should return an array of integers which begins with a number represented by the <code>startNum</code> parameter and ends with a number represented by the <code>endNum</code> parameter.  The starting number will always be less than or equal to the ending number.  Your function must use recursion by calling itself and not use loops of any kind.
```
This just flows a bit better, to my eyes.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]:
```suggestion
  - text: Your code should not use any loop syntax (<code>for</code> or <code>while</code> or higher order functions such as <code>forEach</code>, <code>map</code>, <code>filter</code>, or <code>reduce</code>).
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Looking good!  I noticed a few minor things, but I think it's pretty good.  Initially I was a little worried that the challenge might be too hard, then I realised the solution is a pretty simple extension of the example.  Should be fine!

**Comment by RandellDawson** [comment]
@ojeytonwilliams Thanks for the suggestions.  I have incorporated those into this PR.

**Review by ojeytonwilliams — APPROVED** [review]
Thanks, LGTM!

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (Manish-Giri) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]: I think we need to specify what the expected result should be when both `startNum` and `endNum` are the same. It *is* evident once you calculate recursively, but may not be that clear for beginners.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (Manish-Giri) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]:
```suggestion
  - text: <code>rangeOfNumbers(4, 4)</code> should should return <code>[4]</code>.
```
Currently, shows up as a blank - `rangeOfNumbers(4, 4) should should return .`

**Review by Manish-Giri — CHANGES_REQUESTED** [review]
Some minor suggestions and a fix.

**Review by Manish-Giri — APPROVED** [review]
LGTM

**Comment by RandellDawson** [comment]
@Manish-Giri @scissorsneedfoodtoo It appears we are just waiting on a core team member to approve, so this can be merged.  There is an earlier PR (#35966) which Oliver and I have been discussing offline to add as a 3rd recursion challenge, so this current PR can go ahead and be merged.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]: Not necessary at all, just noticed that the second instance of `startNum` and `endNum` aren't wrapped in code tags:

We have defined a function named <code>rangeOfNumbers</code> with two parameters. The function should return an array of integers which begins with a number represented by the <code>startNum</code> parameter and ends with a number represented by the <code>endNum</code> parameter.  The starting number will always be less than or equal to the ending number.  Your function must use recursion by calling itself and not use loops of any kind. It should also work for cases where both <code>startNum</code> and <code>endNum</code> are the same.

**Review by scissorsneedfoodtoo — APPROVED** [review]
Everything LGTM. Looking forward to seeing this live :+1: :+1:

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md
Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-range-of-numbers.english.md]: @scissorsneedfoodtoo Just added them now.

**Comment by ojeytonwilliams** [comment]
@RandellDawson Thank you for your contribution to the curriculum! :+1:
We're happy to accept this challenge, and look forward to future contributions. :memo:

**Pull request merged**
